File: typescript-workshop/part-3-mocks-clean/src/service/PricingService.ts
import { DiscountCode, discountPercentage } from '../domain/DiscountCode';
import { OrderItem } from '../domain/OrderItem';

const TAX_RATE = 0.20;
const FREE_SHIPPING_THRESHOLD = 100.0;
const SHIPPING_COST = 5.00;

export class PricingService {
  calculateTotal(items: ReadonlyArray<OrderItem>, discountCode?: DiscountCode): number {
    const subtotal = items.reduce((sum, item) => sum + item.quantity * item.unitPrice, 0);
    const discountedSubtotal = discountCode !== undefined
      ? subtotal * (1 - discountPercentage[discountCode] / 100)
      : subtotal;
    const tax = discountedSubtotal * TAX_RATE;
    const shipping = discountedSubtotal >= FREE_SHIPPING_THRESHOLD ? 0 : SHIPPING_COST;
    return discountedSubtotal + tax + shipping;
  }
}

File: csharp-workshop/Part2.LightMocks/Service/PricingService.cs
using Part2.LightMocks.Domain;

namespace Part2.LightMocks.Service;

public class PricingService
{
    public const double TaxRate = 0.20;
    public const double FreeShippingThreshold = 100.0;
    public const double ShippingCost = 5.00;

    public virtual double CalculateTotal(IReadOnlyList<OrderItem> items)
    {
        double subtotal = CalculateSubtotal(items);
        double tax = CalculateTax(subtotal);
        double shipping = CalculateShipping(subtotal);
        return subtotal + tax + shipping;
    }

    public virtual double CalculateTotal(IReadOnlyList<OrderItem> items, DiscountCode discountCode)
    {
        double subtotal = CalculateSubtotal(items);
        double discountedSubtotal = ApplyDiscount(subtotal, discountCode);
        double tax = CalculateTax(discountedSubtotal);
        double shipping = CalculateShipping(discountedSubtotal);
        return discountedSubtotal + tax + shipping;
    }

    private static double CalculateSubtotal(IReadOnlyList<OrderItem> items)
    {
        return items.Sum(item => item.Quantity * item.UnitPrice);
    }

    private static double ApplyDiscount(double subtotal, DiscountCode discountCode)
    {
        double discountFactor = 1.0 - (discountCode.GetPercentage() / 100.0);
        return subtotal * discountFactor;
    }

    private static double CalculateTax(double subtotal)
    {
        return subtotal * TaxRate;
    }

    private static double CalculateShipping(double subtotal)
    {
        return subtotal >= FreeShippingThreshold ? 0.0 : ShippingCost;
    }
}
